neonvm/runner/main.go

```diff
@@ -532,10 +544,198 @@ func main() {
 		qemuCmd = append(qemuCmd, "-incoming", fmt.Sprintf("tcp:0:%d", vmv1.MigrationPort))
 	}
 
-	go terminateQemuOnSigterm(vmSpec.QMP)
-	if err := execFg(QEMU_BIN, qemuCmd...); err != nil {
+	isSystemd, err := isSystemdCgroup()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var cgroupPath string
+	cgroupName := fmt.Sprintf("%s-vm-runner", vmStatus.PodName)
+	if isSystemd {
+		cgroupPath = path.Join(cgroupSystemdBase, cgroupName)
+	} else {
+		cgroupPath = path.Join(cgroupPlainBase, cgroupName)
+	}
+
+	if err := createCgroup(cgroupPath); err != nil {
+		log.Fatal(err)
+	}
+
+	if err := setCgroupLimit(qemuCPUs.minFraction, cgroupPath); err != nil {
 		log.Fatal(err)
 	}
+
+	go terminateQemuOnSigterm(vmSpec.QMP)
+	go listenForCPUChanges(vmSpec.RunnerPort, cgroupPath)
+
+	qemuString := shellescape.QuoteCommand(append([]string{QEMU_BIN}, qemuCmd...))
+	cgropedQemu := fmt.Sprintf("echo $$ >> %s && %s", path.Join(cgroupPath, "cgroup.procs"), qemuString)
+	// start a new shell, cgroup it and then start qemu. Qemu will get cgroup from the parent sh while runner is not cgrouped
+	if err := execFg("sh", "-c", cgropedQemu); err != nil {
+		log.Println(err)
```

Comment:
why `log` and not `klog`

Reply:
for context: we use klog (almost) everywhere else

Reply:
Makes sense, good to leave as-is for now. Opened #218 to track this.